# Reconstruct UKBB std data

In [35]:
import os
import pandas as pd


In [36]:
## use full function

## check categorical variable frequence for each level 
def freq_chk(dataframe,var='',output=True):
    ctab=pd.crosstab(dataframe[var], columns='count')
    if output==True:
        print(ctab)
    else:
        pass
    return ctab;

## load variable description Table

In [37]:
os.chdir("/temp_project/ukbb/data")
df_tab1_adv=pd.read_csv('ukb22598_html_tab1_adv.csv')
df_tab1_adv['DC'] = df_tab1_adv['DC'].fillna('NaN')
df_tab1_adv['CatLv'] = df_tab1_adv['CatLv'].fillna('NaN')
df_tab1_adv.head()


Column     UDI   Count                  Type  \
0       0     eid  502602              Sequence   
1       1  21-0.0  500887  Categorical (single)   
2       2  21-1.0   20336  Categorical (single)   
3       3  21-2.0   23167  Categorical (single)   
4       4  31-0.0  502602  Categorical (single)   

                                         Description  Tier      DC CatLv  \
0                  Encoded anonymised participant ID     1     NaN   NaN   
1  Weight methodUses data-coding 100261 comprises...     1  100261     5   
2  Weight methodUses data-coding 100261 comprises...     1  100261     5   
3  Weight methodUses data-coding 100261 comprises...     1  100261     5   
4  SexUses data-coding 9 comprises 2 Integer-valu...     1       9     2   

  udip1 udip2  udip2m  udip2n  miss_ct   miss_pt  obs_ct    obs_pt  
0   eid   eid    -999    -999        0  0.000000  502602  1.000000  
1    21   0.0       0       0     1715  0.003412  500887  0.996588  
2    21   1.0       1       0   482266  0.959539   20336  0.040461  
3    21   2.0       2       0   479435  0.953906   23167  0.046094  
4    31   0.0       0       0        0  0.000000  502602  1.000000

## variable of different types

In [38]:
vartab_cam = df_tab1_adv[df_tab1_adv.Type.str.contains("multiple")==True]
vartab_cas = df_tab1_adv[df_tab1_adv.Type.str.contains("single")==True]
vartab_con = df_tab1_adv[df_tab1_adv.Type.str.contains("Continuous")==True]
vartab_cur = df_tab1_adv[df_tab1_adv.Type.str.contains("Curve")==True]
vartab_dat = df_tab1_adv[df_tab1_adv.Type.str.contains("Date")==True]
vartab_int = df_tab1_adv[df_tab1_adv.Type.str.contains("Integer")==True]
vartab_tex = df_tab1_adv[df_tab1_adv.Type.str.contains("Text")==True]
vartab_tim = df_tab1_adv[df_tab1_adv.Type.str.contains("Time")==True]

print('CAT MU:',vartab_cam.udip1.count()) 
print('CAT SI:',vartab_cas.udip1.count()) 
print('CON:   ',vartab_con.udip1.count()) 
print('CUR:   ',vartab_cur.udip1.count()) 
print('DAT:   ',vartab_dat.udip1.count()) 
print('INT:   ',vartab_int.udip1.count()) 
print('TEX:   ',vartab_tex.udip1.count()) 
print('TIM:   ',vartab_tim.udip1.count()) 

CAT MU: 2300
CAT SI: 2563
CON:    910
CUR:    5
DAT:    39
INT:    2309
TEX:    623
TIM:    27


## unique variable of different types

In [39]:
vartab_cam_uniq = vartab_cam.groupby('udip1').head(1).reset_index(drop=True)
vartab_cas_uniq = vartab_cas.groupby('udip1').head(1).reset_index(drop=True)
vartab_con_uniq = vartab_con.groupby('udip1').head(1).reset_index(drop=True)
vartab_cur_uniq = vartab_cur.groupby('udip1').head(1).reset_index(drop=True)
vartab_dat_uniq = vartab_dat.groupby('udip1').head(1).reset_index(drop=True)
vartab_int_uniq = vartab_int.groupby('udip1').head(1).reset_index(drop=True)
vartab_tex_uniq = vartab_tex.groupby('udip1').head(1).reset_index(drop=True)
vartab_tim_uniq = vartab_tim.groupby('udip1').head(1).reset_index(drop=True)
print('CAT MU:',vartab_cam_uniq.udip1.count()) 
print('CAT SI:',vartab_cas_uniq.udip1.count()) 
print('CON:   ',vartab_con_uniq.udip1.count()) 
print('CUR:   ',vartab_cur_uniq.udip1.count()) 
print('DAT:   ',vartab_dat_uniq.udip1.count()) 
print('INT:   ',vartab_int_uniq.udip1.count()) 
print('TEX:   ',vartab_tex_uniq.udip1.count()) 
print('TIM:   ',vartab_tim_uniq.udip1.count()) 

CAT MU: 56
CAT SI: 504
CON:    129
CUR:    3
DAT:    4
INT:    257
TEX:    72
TIM:    7


In [55]:
56+504+129+3+4+257+72+7+1

1033


## varible type summary: 

| Type                   | count* | unique count |
|------------------------|-------|---------------|
| Categorical (multiple) | 2300  |56|
| Categorical (single)   | 2563  |504|
| Continuous             | 910   |129|
| Curve                  | 5     |3|
| Date                   | 39    |4|
| Integer                | 2309  |257|
| Sequence**               | 1     |1|
| Text                   | 623   |72|
| Time                   | 27    |7|
|------------------------|-------|---------------|
| sum                     |8777   |1033           |
*count is before reconstruction

**'eid' is the sequence type 



In [40]:
print('unique variable count: '+ str(len(df_tab1_adv.udip1.unique())))

unique variable count: 1033


## check instance type

In [41]:
### dataframe contains all variables that at instance=0
uniqchk=df_tab1_adv.groupby('udip1').head(1).reset_index(drop=True)

uniqchk.udip2m.value_counts()

 0      1021
 2        11
-999       1
Name: udip2m, dtype: int64

In [42]:
#uniqchk[uniqchk['udip2m']==2]  ## start from -2.x has 11 unique counts, in total 26 variables

uniqchk_sta_not0x=uniqchk[uniqchk['udip2m']!=0] 
list_uniqchk_sta_not0x=uniqchk_sta_not0x['udip1'].tolist()
list_uniqchk_sta_not0x       ## eid + 11 unique var starts -2.x, that is unique 12 vars, in total 27 variable

uniqchk[uniqchk['udip2m']==0].udip2n.value_counts()   
## start from -0.x has 1021 unique counts (in total 8750 variables), where start as 
## -0.0 has 1001
## -0.1 has 20

uniqchk_sta0x=uniqchk[uniqchk['udip2m']==0]   ## 1021 unique var starts like '-0.x'
uniqchk_sta01=uniqchk_sta0x[uniqchk_sta0x['udip2n']==1] ## 20 unique var starts like '-0.1'
uniqchk_sta00=uniqchk_sta0x[uniqchk_sta0x['udip2n']==0] ## 1001 unique var starts like '-0.0'

list_uniqchk_sta01=uniqchk_sta01['udip1'].tolist()
list_uniqchk_sta00=uniqchk_sta00['udip1'].tolist()

df_stall=df_tab1_adv[~df_tab1_adv.udip1.isin(list_uniqchk_sta_not0x)]    # len=8570 = 8777-27


df_sta01=df_stall[df_stall.udip1.isin(list_uniqchk_sta01)]
df_sta00=df_stall[df_stall.udip1.isin(list_uniqchk_sta00)]
df_sta2x=df_tab1_adv[df_tab1_adv.udip1.isin(list_uniqchk_sta_not0x)]

#len(df_sta01)   ## 232 vars
#len(df_sta00)   ## 8518 vars
#len(df_sta2x)   ## 27 vars

df_sta01_for_icount=df_sta01[df_sta01['udip2n']==1]
df_sta00_for_icount=df_sta00[df_sta00['udip2n']==0]

#len(df_sta01_for_icount)   ## 30
#len(df_sta00_for_icount)   ## 2529

In [43]:
### Add instance information

In [44]:
field_i3or5_sta01=df_sta01_for_icount.udip1.value_counts()
field_i3or5_sta01.to_csv('/temp_project/ukbb/data/i0/varInstanceInfo/afield_i3or5_sta01.csv')

field_i3or5_sta00=df_sta00_for_icount.udip1.value_counts()
field_i3or5_sta00.to_csv('/temp_project/ukbb/data/i0/varInstanceInfo/afield_i3or5_sta00.csv')

field_i3or5_sta01=pd.read_csv('/temp_project/ukbb/data/i0/varInstanceInfo/afield_i3or5_sta01.csv',header=None)
field_i3or5_sta01.columns=['field','instance_type']

field_i3or5_sta00=pd.read_csv('/temp_project/ukbb/data/i0/varInstanceInfo/afield_i3or5_sta00.csv',header=None)
field_i3or5_sta00.columns=['field','instance_type']

field_i3or5_sta01
field_i3or5_to_create_ind=pd.merge(field_i3or5_sta00, field_i3or5_sta01, on="field", how='outer')
field_i3or5_to_create_ind['instance_type']=field_i3or5_to_create_ind.drop('field',axis=1).max(axis=1)
#len(field_i3or5_sta01)   ## 20
#len(field_i3or5_sta00)   ## 1001
#len(field_i3or5_to_create_ind)   ## 1021

field_i3or5_to_ind_pre=pd.get_dummies(field_i3or5_to_create_ind, prefix='instance_type_ind', columns=['instance_type'])
field_i3or5_to_ind=field_i3or5_to_ind_pre[['field','instance_type_ind_3.0','instance_type_ind_5.0','instance_type_ind_30.0','instance_type_ind_32.0']]
field_i3or5_to_ind.columns=['udip1','instance_type_ind3','instance_type_ind5','instance_type_ind30','instance_type_ind32']
field_i3or5_to_ind.head()


udip1  instance_type_ind3  instance_type_ind5  instance_type_ind30  \
0  40006                   0                   0                    0   
1  40008                   0                   0                    0   
2  40005                   0                   0                    0   
3  40012                   0                   0                    0   
4  40011                   0                   0                    0   

   instance_type_ind32  
0                    1  
1                    1  
2                    1  
3                    1  
4                    1

In [45]:
pwd()

'/temp_project/ukbb/data'

In [46]:
### Get comphensive table 1
### UDI restructure varaibles
### intance type variables

df1=df_tab1_adv.copy()
df2=field_i3or5_to_ind

dtype = dict(udip1=str)
#df_tab1_adv2=df1.astype(dtype).merge(df2.astype(dtype), 'left')
df_tab1_adv2=pd.merge(df1.astype(dtype),df2.astype(dtype), on='udip1', how='left')
df_tab1_adv2.to_csv('ukb22598_html_tab1_comp.csv',index=False)

In [47]:
## variables at i0

In [64]:
os.chdir("/temp_project/ukbb/data/i0/varlist")
### All unique variables: 

vartab_i0=df_tab1_adv[df_tab1_adv.udip2m==0]
vartab_i0_uniq=vartab_i0.groupby('udip1').head(1).reset_index(drop=True)

vartab_i0_uniq_for_merge=vartab_i0_uniq[['Type','Description','Tier','DC','CatLv','udip1']]
vartab_i0_uniq_for_merge.Tier.value_counts()

1    440
3    429
2     80
4     72
Name: Tier, dtype: int64


At instance=0

|Type|Measure Times|count|store|
|---|---|
|CON|single|118|as-is|
|   |multiple|11 | as vector|
|CUR|single|3|as-is|
|   |multiple| 0 | NA|
|DAT|single|4|as-is|
|   |multiple|0 | NA|
|INT|single|206|as-is|
|   |multiple|45 | as vector|
|TEX|single|63|as-is|
|   |multiple| 5| as vector|
|TIM|single|6|as-is|
|   |multiple|1 | as vector|
|CAS|single|469|as-is|
|   |multiple|34 | as vector|
|CAM|single|0|NA|
|   |multiple|56 | as vector|

|Tier| Count|
|---|
|1|    440|
|2|     80|
|3|    429|
|4|     72|

## All var at i=0, save the fid by variable type

In [65]:
os.chdir("/temp_project/ukbb/data/i0/varlist")
### All con variables: 
vartab_i0_con=vartab_con[vartab_con['udip2m']==0]
field_i0_con=vartab_i0_con.udip1.value_counts()
field_i0_con.to_csv('afield_i0_con.csv')
### All cur variables: 
vartab_i0_cur=vartab_cur[vartab_cur['udip2m']==0]
field_i0_cur=vartab_i0_cur.udip1.value_counts()
field_i0_cur.to_csv('afield_i0_cur.csv')
### All dat variables: 
vartab_i0_dat=vartab_dat[vartab_dat['udip2m']==0]
field_i0_dat=vartab_i0_dat.udip1.value_counts()
field_i0_dat.to_csv('afield_i0_dat.csv')
### All int variables: 
vartab_i0_int=vartab_int[vartab_int['udip2m']==0]
field_i0_int=vartab_i0_int.udip1.value_counts()
field_i0_int.to_csv('afield_i0_int.csv')
### All tex variables: 
vartab_i0_tex=vartab_tex[vartab_tex['udip2m']==0]
field_i0_tex=vartab_i0_tex.udip1.value_counts()
field_i0_tex.to_csv('afield_i0_tex.csv')
### All tim variables: 
vartab_i0_tim=vartab_tim[vartab_tim['udip2m']==0]
field_i0_tim=vartab_i0_tim.udip1.value_counts()
field_i0_tim.to_csv('afield_i0_tim.csv')
### All cas variables: 
vartab_i0_cas=vartab_cas[vartab_cas['udip2m']==0]
field_i0_cas=vartab_i0_cas.udip1.value_counts()
field_i0_cas.to_csv('afield_i0_cas.csv')
### All cam variables: 
vartab_i0_cam=vartab_cam[vartab_cam['udip2m']==0]
field_i0_cam=vartab_i0_cam.udip1.value_counts()
field_i0_cam.to_csv('afield_i0_cam.csv')

## Unique var at i=0

In [66]:
vartab_i0_cam_uniq = vartab_i0_cam.groupby('udip1').head(1).reset_index(drop=True)
vartab_i0_cas_uniq = vartab_i0_cas.groupby('udip1').head(1).reset_index(drop=True)
vartab_i0_con_uniq = vartab_i0_con.groupby('udip1').head(1).reset_index(drop=True)
vartab_i0_cur_uniq = vartab_i0_cur.groupby('udip1').head(1).reset_index(drop=True)
vartab_i0_dat_uniq = vartab_i0_dat.groupby('udip1').head(1).reset_index(drop=True)
vartab_i0_int_uniq = vartab_i0_int.groupby('udip1').head(1).reset_index(drop=True)
vartab_i0_tex_uniq = vartab_i0_tex.groupby('udip1').head(1).reset_index(drop=True)
vartab_i0_tim_uniq = vartab_i0_tim.groupby('udip1').head(1).reset_index(drop=True)
print('CAT MU:',vartab_i0_cam_uniq.udip1.count()) 
print('CAT SI:',vartab_i0_cas_uniq.udip1.count()) 
print('CON:   ',vartab_i0_con_uniq.udip1.count()) 
print('CUR:   ',vartab_i0_cur_uniq.udip1.count()) 
print('DAT:   ',vartab_i0_dat_uniq.udip1.count()) 
print('INT:   ',vartab_i0_int_uniq.udip1.count()) 
print('TEX:   ',vartab_i0_tex_uniq.udip1.count()) 
print('TIM:   ',vartab_i0_tim_uniq.udip1.count()) 

CAT MU: 56
CAT SI: 503
CON:    129
CUR:    3
DAT:    4
INT:    251
TEX:    68
TIM:    7


In [67]:
56+503+129+3+4+251+68+7+1

1022


## varible type summary: 

| Type                   | count* | unique count |unique count at i0|
|------------------------|-------|---------------|
| Categorical (multiple) | 2300  |56|56|
| Categorical (single)   | 2563  |504|503|
| Continuous             | 910   |129|129|
| Curve                  | 5     |3|3|
| Date                   | 39    |4|4|
| Integer                | 2309  |257|251|
| Sequence**               | 1     |1|1|
| Text                   | 623   |72|68
| Time                   | 27    |7|7|
|------------------------|-------|--------|-------|
| sum                     |8777   |1033    | 1022      |
*count is before reconstruction

**'eid' is the sequence type 



In [ ]:
### read file with vector in variable, which is the file name starts with a 'vec_'
'''
pd.read_csv('.../...csv', sep='\t')
'''
### read file without vector in variable
'''
pd.read_csv('.../...csv')
'''


df_tab1_adv[df_tab1_adv.Type=='Categorical (multiple)']
df_tab1_adv[df_tab1_adv.udip1=='6142']
df_tab1_adv[(df_tab1_adv.Type=='Continuous')]
df_tab1_adv[df_tab1_adv.udip1=='41204']

## Below is the code to restructure i0 data and save them into corresponding folder
## Need to run it on SDSC serve.
## Don't run them here

## At i=0, single measure keep as-is, multiple measures keep as a vector

In [ ]:
os.chdir("/temp_project/ukbb/data/i0")

In [ ]:
var_type_list=['con','cur','dat','int','tex','tim','cas','cam']

### Var CON

In [ ]:

os.chdir("/temp_project/ukbb/data/i0/varlist")
### AS-IS for single measure con variable

fid_i0_con=pd.read_csv('afield_i0_con.csv',header=None)
fid_i0_con.columns=['field','count']
#fid_i0_con.loc.field=fid_i0_con.field.astype(str)



### At i=0, lsit of continuous var only 1 measurement
list_i0_con_1=fid_i0_con[fid_i0_con['count']==1].field.astype(str).tolist()
UDI_i0_con_1=vartab_i0_con.loc[vartab_i0_con['udip1'].isin(list_i0_con_1)]
UDI_list_i0_con_1=UDI_i0_con_1['UDI'].tolist()
UDI_i0_con_1['UDI'].to_csv('aUDI_i0_con_1.csv',header='UDI')

#UDI_i0_con_1=pd.read_csv('aUDI_i0_con_1.csv')
#UDI_i0_con_1
#UDI_list_i0_con_1=UDI_i0_con_1['UDI'].tolist()




In [ ]:
os.chdir("/temp_project/ukbb/data/i0/varlist")
### Vector for multi-measures con variable

fid_i0_con=pd.read_csv('afield_i0_con.csv',header=None)
fid_i0_con.columns=['field','count']
#fid_i0_con.loc.field=fid_i0_con.field.astype(str)

### At i=0, lsit of continuous var >= 2 measurements
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_con_2=fid_i0_con[fid_i0_con['count']>1].field.astype(str).tolist()

#UDI_i0_con_2=vartab_i0_con.loc[vartab_i0_con['udip1'].isin(list_i0_con_2)]
#UDI_list_i0_con_1=UDI_i0_con_1['UDI'].tolist()
#UDI_i0_con_1['UDI'].to_csv('aUDI_i0_con_1.csv',header='UDI')


list_i0_con_2part=list_i0_con_2[0:3]

for fid in list_i0_con_2part:
    
    UDI_i0_con_2=vartab_i0_con.loc[vartab_i0_con['udip1'].isin([fid])]
    UDI_list_i0_con_2=UDI_i0_con_2['UDI'].tolist()
    
    loadcol=['eid']+UDI_list_i0_con_2
    colname='vec_con'+str(fid)+'i0'
    
    ofile=pd.read_csv('/temp_project/ukbb/ukb22598.csv', usecols=loadcol, nrows=502602,low_memory=False)
    ofile[str(colname)] = ofile[ofile.columns[1:]].apply(lambda x: ','.join(x.dropna().astype(float).astype(str)),axis=1)
    
    vec_df=ofile[['eid',str(colname)]]
    outfilename='/temp_project/ukbb/data/i0/var_con/'+str(colname)+'.csv'
    vec_df.to_csv(outfilename, index=False, sep='\t')
    

### Var INT

In [ ]:

os.chdir("/temp_project/ukbb/data/i0/varlist")
### AS-IS for single measure INT variable

fid_i0_int=pd.read_csv('afield_i0_int.csv',header=None)
fid_i0_int.columns=['field','count']
#fid_i0_int.loc.field=fid_i0_int.field.astype(str)



### At i=0, lsit of INT var only 1 measurement
list_i0_int_1=fid_i0_int[fid_i0_int['count']==1].field.astype(str).tolist()
UDI_i0_int_1=vartab_i0_int.loc[vartab_i0_int['udip1'].isin(list_i0_int_1)]
UDI_list_i0_int_1=UDI_i0_int_1['UDI'].tolist()
UDI_i0_int_1['UDI'].to_csv('aUDI_i0_int_1.csv',header='UDI')

#len(list_i0_int_1)
#206

#UDI_i0_int_1=pd.read_csv('aUDI_i0_int_1.csv')
#UDI_i0_int_1
#UDI_list_i0_int_1=UDI_i0_int_1['UDI'].tolist()


In [ ]:
os.chdir("/temp_project/ukbb/data/i0/varlist")
### Vector for multi-measures INT variable

fid_i0_int=pd.read_csv('afield_i0_int.csv',header=None)
fid_i0_int.columns=['field','count']
#fid_i0_int.loc.field=fid_i0_int.field.astype(str)

### At i=0, lsit of INT var >= 2 measurements
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_int_2=fid_i0_int[fid_i0_int['count']>1].field.astype(str).tolist()

#UDI_i0_int_2=vartab_i0_int.loc[vartab_i0_int['udip1'].isin(list_i0_int_2)]
#UDI_list_i0_int_1=UDI_i0_int_1['UDI'].tolist()
#UDI_i0_int_1['UDI'].to_csv('aUDI_i0_int_1.csv',header='UDI')


list_i0_int_2part=list_i0_int_2[0:3]

for fid in list_i0_int_2part:
    
    UDI_i0_int_2=vartab_i0_int.loc[vartab_i0_int['udip1'].isin([fid])]
    UDI_list_i0_int_2=UDI_i0_int_2['UDI'].tolist()
    
    loadcol=['eid']+UDI_list_i0_int_2
    colname='vec_int'+str(fid)+'i0'
    
    ofile=pd.read_csv('/temp_project/ukbb/ukb22598.csv', usecols=loadcol, nrows=502602,low_memory=False)
    ofile[str(colname)] = ofile[ofile.columns[1:]].apply(lambda x: ','.join(x.dropna().astype(int).astype(str)),axis=1)
    
    vec_df=ofile[['eid',str(colname)]]
    outfilename='/temp_project/ukbb/data/i0/var_int/'+str(colname)+'.csv'
    vec_df.to_csv(outfilename, index=False, sep='\t')

### Var DAT

### Var TIM

In [ ]:
os.chdir("/temp_project/ukbb/data/i0/varlist")
### AS-IS for single measure TIM variable

fid_i0_tim=pd.read_csv('afield_i0_tim.csv',header=None)
fid_i0_tim.columns=['field','count']

### At i=0, lsit of TIM var only 1 measurement
list_i0_tim_1=fid_i0_tim[fid_i0_tim['count']==1].field.astype(str).tolist()
UDI_i0_tim_1=vartab_i0_tim.loc[vartab_i0_tim['udip1'].isin(list_i0_tim_1)]
UDI_list_i0_tim_1=UDI_i0_tim_1['UDI'].tolist()
UDI_i0_tim_1['UDI'].to_csv('aUDI_i0_tim_1.csv',header='UDI')

#len(list_i0_tim_1)
#6

In [ ]:
os.chdir("/temp_project/ukbb/data/i0/varlist")
### Vector for multi-measures TIM variable

fid_i0_tim=pd.read_csv('afield_i0_tim.csv',header=None)
fid_i0_tim.columns=['field','count']
#fid_i0_tim.loc.field=fid_i0_tim.field.astype(str)

### At i=0, lsit of TIM var >= 2 measurements
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_tim_2=fid_i0_tim[fid_i0_tim['count']>1].field.astype(str).tolist()

#len(list_i0_tim_2)
#1


for fid in list_i0_tim_2:
    
    UDI_i0_tim_2=vartab_i0_tim.loc[vartab_i0_tim['udip1'].isin([fid])]
    UDI_list_i0_tim_2=UDI_i0_tim_2['UDI'].tolist()
    
    loadcol=['eid']+UDI_list_i0_tim_2
    colname='vec_tim'+str(fid)+'i0'
    
    ofile=pd.read_csv('/temp_project/ukbb/ukb22598.csv', usecols=loadcol, nrows=502602,low_memory=False)
    ofile[str(colname)] = ofile[ofile.columns[1:]].apply(lambda x: ','.join(x.dropna().astype(str)),axis=1)
    
    vec_df=ofile[['eid',str(colname)]]
    outfilename='/temp_project/ukbb/data/i0/var_tim/'+str(colname)+'.csv'
    vec_df.to_csv(outfilename, index=False, sep='\t')

### Var CUR

### Var TEX

In [ ]:


os.chdir("/temp_project/ukbb/data/i0/varlist")
### AS-IS for single measure TEX variable

fid_i0_tex=pd.read_csv('afield_i0_tex.csv',header=None)
fid_i0_tex.columns=['field','count']

### At i=0, lsit of TEX var only 1 measurement
list_i0_tex_1=fid_i0_tex[fid_i0_tex['count']==1].field.astype(str).tolist()
UDI_i0_tex_1=vartab_i0_tex.loc[vartab_i0_tex['udip1'].isin(list_i0_tex_1)]
UDI_list_i0_tex_1=UDI_i0_tex_1['UDI'].tolist()
UDI_i0_tex_1['UDI'].to_csv('aUDI_i0_tex_1.csv',header='UDI')

#len(list_i0_tex_1)
#63


In [ ]:
# TEX example
#UDI_i0_tex_2=vartab_i0_tex.loc[vartab_i0_tex['udip1'].isin(list_i0_tex_2)]
#UDI_i0_tex_2
#UDI_i0_tex_1

In [ ]:

os.chdir("/temp_project/ukbb/data/i0/varlist")
### Vector for multi-measures TEX variable

fid_i0_tex=pd.read_csv('afield_i0_tex.csv',header=None)
fid_i0_tex.columns=['field','count']
#fid_i0_tex.loc.field=fid_i0_tex.field.astype(str)

### At i=0, lsit of TEX var >= 2 measurements
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_tex_2=fid_i0_tex[fid_i0_tex['count']>1].field.astype(str).tolist()

#len(list_i0_tex_2)
#5

for fid in list_i0_tex_2:
    
    UDI_i0_tex_2=vartab_i0_tex.loc[vartab_i0_tex['udip1'].isin([fid])]
    UDI_list_i0_tex_2=UDI_i0_tex_2['UDI'].tolist()
    
    loadcol=['eid']+UDI_list_i0_tex_2
    colname='vec_tex'+str(fid)+'i0'
    
    ofile=pd.read_csv('/temp_project/ukbb/ukb22598.csv', usecols=loadcol, nrows=502602,low_memory=False)
    ofile[str(colname)] = ofile[ofile.columns[1:]].apply(lambda x: ','.join(x.dropna().astype(str)),axis=1)
    
    vec_df=ofile[['eid',str(colname)]]
    outfilename='/temp_project/ukbb/data/i0/var_tex/'+str(colname)+'.csv'
    vec_df.to_csv(outfilename, index=False, sep='\t')

### Var CAS

In [ ]:
os.chdir("/temp_project/ukbb/data/i0/varlist")
### AS-IS for single measure CAS variable

fid_i0_cas=pd.read_csv('afield_i0_cas.csv',header=None)
fid_i0_cas.columns=['field','count']

### At i=0, lsit of CAS var only 1 measurement
list_i0_cas_1=fid_i0_cas[fid_i0_cas['count']==1].field.astype(str).tolist()
UDI_i0_cas_1=vartab_i0_cas.loc[vartab_i0_cas['udip1'].isin(list_i0_cas_1)]
UDI_list_i0_cas_1=UDI_i0_cas_1['UDI'].tolist()
UDI_i0_cas_1['UDI'].to_csv('aUDI_i0_cas_1.csv',header='UDI')

#len(list_i0_cas_1)
#469



In [ ]:
os.chdir("/temp_project/ukbb/data/i0/varlist")
### Vector for multi-measures CAS variable

fid_i0_cas=pd.read_csv('afield_i0_cas.csv',header=None)
fid_i0_cas.columns=['field','count']
#fid_i0_cas.loc.field=fid_i0_cas.field.astype(str)

### At i=0, lsit of CAS var >= 2 measurements
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_cas_2=fid_i0_cas[fid_i0_cas['count']>1].field.astype(str).tolist()

#len(list_i0_cas_2)
#34


list_i0_cas_2part=list_i0_cas_2[0:11]

for fid in list_i0_cas_2part:
    
    UDI_i0_cas_2=vartab_i0_cas.loc[vartab_i0_cas['udip1'].isin([fid])]
    UDI_list_i0_cas_2=UDI_i0_cas_2['UDI'].tolist()
    
    loadcol=['eid']+UDI_list_i0_cas_2
    colname='vec_cas'+str(fid)+'i0'
    
    ofile=pd.read_csv('/temp_project/ukbb/ukb22598.csv', usecols=loadcol, nrows=502602,low_memory=False)
    ofile[str(colname)] = ofile[ofile.columns[1:]].apply(lambda x: ','.join(x.dropna().astype(int).astype(str)),axis=1)
    
    vec_df=ofile[['eid',str(colname)]]
    outfilename='/temp_project/ukbb/data/i0/var_cas/'+str(colname)+'.csv'
    vec_df.to_csv(outfilename, index=False, sep='\t')

### Var CAM

In [ ]:
os.chdir("/temp_project/ukbb/data/i0/varlist")
### AS-IS for single measure CAM variable

fid_i0_cam=pd.read_csv('afield_i0_cam.csv',header=None)
fid_i0_cam.columns=['field','count']

### At i=0, lsit of CAM var only 1 measurement
list_i0_cam_1=fid_i0_cam[fid_i0_cam['count']==1].field.astype(str).tolist()
UDI_i0_cam_1=vartab_i0_cam.loc[vartab_i0_cam['udip1'].isin(list_i0_cam_1)]
UDI_list_i0_cam_1=UDI_i0_cam_1['UDI'].tolist()
UDI_i0_cam_1['UDI'].to_csv('aUDI_i0_cam_1.csv',header='UDI')

#len(list_i0_cam_1)
#NaN


In [ ]:
os.chdir("/temp_project/ukbb/data/i0/varlist")
### Vector for multi-measures CAM variable

fid_i0_cam=pd.read_csv('afield_i0_cam.csv',header=None)
fid_i0_cam.columns=['field','count']
#fid_i0_cam.loc.field=fid_i0_cam.field.astype(str)

### At i=0, lsit of CAM var >= 2 measurements
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_cam_2=fid_i0_cam[fid_i0_cam['count']>1].field.astype(str).tolist()
#len(list_i0_cam_2)
#56

#UDI_i0_cam_2=vartab_i0_cam.loc[vartab_i0_cam['udip1'].isin(list_i0_cam_2)]
#UDI_list_i0_cam_1=UDI_i0_cam_1['UDI'].tolist()
#UDI_i0_cam_1['UDI'].to_csv('aUDI_i0_cam_1.csv',header='UDI')


list_i0_cam_2part=list_i0_cam_2[0:15]

for fid in list_i0_cam_2part:
    
    UDI_i0_cam_2=vartab_i0_cam.loc[vartab_i0_cam['udip1'].isin([fid])]
    UDI_list_i0_cam_2=UDI_i0_cam_2['UDI'].tolist()
    
    loadcol=['eid']+UDI_list_i0_cam_2
    colname='vec_cam'+str(fid)+'i0'
    
    ofile=pd.read_csv('/temp_project/ukbb/ukb22598.csv', usecols=loadcol, nrows=502602,low_memory=False)
    ofile[str(colname)] = ofile[ofile.columns[1:]].apply(lambda x: ','.join(x.dropna().astype(int).astype(str)),axis=1)
    
    vec_df=ofile[['eid',str(colname)]]
    outfilename='/temp_project/ukbb/data/i0/var_cam/'+str(colname)+'.csv'
    vec_df.to_csv(outfilename, index=False, sep='\t')

In [ ]:
list_i0_con_2

In [ ]:
rt=[list_i0_con_2[9]]
UDI_i0_con_2=vartab_i0_con.loc[vartab_i0_con['udip1'].isin(rt)]
UDI_list_i0_con_2=UDI_i0_con_2['UDI'].tolist()
r=UDI_list_i0_con_2

loadcol=['eid']+r
loadcol

ofile=pd.read_csv('/temp_project/ukbb/ukb22598.csv', usecols=loadcol, nrows=502602)
ofile[str(rt[0])] = ofile[ofile.columns[1:]].apply(lambda x: ','.join(x.dropna().astype(float).astype(str)),axis=1)
ofile

In [ ]:
str(rt)

In [ ]:
ofile=pd.read_csv('/temp_project/ukbb/ukb22598.csv', usecols=loadcol, nrows=502602)

In [ ]:
ofile.columns[1:]
import numpy as np
cls=df['20006'][3].split(',')

In [ ]:
cls

In [ ]:
df=ofile[['eid','20006']]
df.head()
#df['avg']=df[df['20006'].apply(np.mean,axis=1)]
results =[float(i) for i in cls]
np.mean(results)


In [ ]:
df['avg']=df[df['20006']].apply(lambda x: np.mean(x.dropna().astype(float).split(',')),axis=1)

In [ ]:
#df['avg']=df['20006'].apply(lambda x: x.split(','))
#df

df['avg']=df[df['20006']].apply(lambda x: np.mean([float(i) for i in x.split(',')]),axis=1)

In [ ]:
ofile[str(rt[0])]=ofile['ColumnA'] 
ofile.head()

In [ ]:
#ofile['ColumnA'] = ofile[ofile.columns[1:]].apply(lambda x: ','.join(x.dropna().astype(float).astype(str)),axis=1)
#ofile

In [ ]:
### All continuous variables: 
vartab_cont_i0=vartab_cont[vartab_cont['udip2m']==0]
tolist=freq_chk(vartab_cont_i0,'udip1',output=False)
#vartab_cont_uniq

### At i=0, lsit of continuous var only 1 measurement
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_con_1=tolist[tolist['count']==1].index.tolist()

### At i=0, lsit of continuous var >= 2 measurements
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_con_2=tolist[tolist['count']>1].index.tolist()

list_i0_con_1

UDI_i0_con_1=vartab_cont_i0.loc[vartab_cont_i0['udip1'].isin(list_i0_con_1)]
#UDI_i0_con_1
UDI_list_i0_con_1=UDI_i0_con_1['UDI'].tolist()
#for r in list_i0_con_1:
 #   vartab_cont_i0
#UDI_list_i0_con_1
UDI_i0_con_1['UDI'].to_csv('aUDI_i0_con_1.csv',header='UDI')

UDI_i0_con_1=pd.read_csv('aUDI_i0_con_1.csv')
UDI_i0_con_1
UDI_list_i0_con_1=UDI_i0_con_1['UDI'].tolist()
#UDI_list_i0_con_1


#list_i0_con_2   ###### multiple measures to a vector
UDI_i0_con_2=vartab_cont_i0.loc[vartab_cont_i0['udip1'].isin(list_i0_con_1)]
#UDI_i0_con_1




In [ ]:
con_2=vartab_cont_i0.udip1.value_counts()

con_2.to_csv('con_2.csv')

con_22=pd.read_csv('con_2.csv',header=None)
con_22.columns=['field','conut']
con_22

In [ ]:
#import numpy as np
#np.savetxt("UDI_list_i0_con_1.csv", UDI_list_i0_con_1, delimiter=",", fmt='%s', header='i0con')

In [ ]:

fid_i0_int=pd.read_csv('afield_i0_int.csv',header=None)
fid_i0_int.columns=['field','count']
#fid_i0_int.loc.field=fid_i0_int.field.astype(str)

### At i=0, lsit of INT var >= 2 measurements
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_int_2=fid_i0_int[fid_i0_int['count']>1].field.astype(str).tolist()
len(list_i0_int_2)

In [ ]:






vartab_con = df_tab1_adv[df_tab1_adv.Type.str.contains("Continuous")==True]
vartab_i0_con=vartab_con[vartab_con['udip2m']==0]




     

In [ ]:
### All integer variables: 


os.chdir("/temp_project/ukbb/data/i0")

vartab_int_i0=vartab_int[vartab_int['udip2m']==0]
to_int_list=freq_chk(vartab_int_i0,'udip1',output=False)

### At i=0, lsit of integer var only 1 measurement
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_int_1=to_int_list[to_int_list['count']==1].index.tolist()

### At i=0, lsit of integer var >= 2 measurements
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_int_2=to_int_list[to_int_list['count']>1].index.tolist()
vartab_int_i0.head()



list_i0_int_1
#list_i0_int_2   ###### Need to discuss about how to re-structure

UDI_i0_int_1=vartab_int_i0.loc[vartab_int_i0['udip1'].isin(list_i0_int_1)]
UDI_list_i0_int_1=UDI_i0_int_1['UDI'].tolist()
#for r in list_i0_int_1:
 #   vartab_int_i0
#UDI_list_i0_int_1
UDI_i0_int_1['UDI'].to_csv('aUDI_i0_int_1.csv',header='UDI')

UDI_i0_int_1=pd.read_csv('aUDI_i0_int_1.csv')
UDI_i0_int_1
UDI_list_i0_int_1=UDI_i0_int_1['UDI'].tolist()
#UDI_list_i0_int_1



In [ ]:
### All date variables: 


os.chdir("/temp_project/ukbb/data/i0")

vartab_dat_i0=vartab_date[vartab_date['udip2m']==0]
to_dat_list=freq_chk(vartab_dat_i0,'udip1',output=False)

### At i=0, lsit of date var only 1 measurement
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_dat_1=to_dat_list[to_dat_list['count']==1].index.tolist()

### At i=0, lsit of date var >= 2 measurements
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_dat_2=to_dat_list[to_dat_list['count']>1].index.tolist()
vartab_dat_i0.head()



list_i0_dat_1
#list_i0_dat_2   ###### Need to discuss about how to re-structure

UDI_i0_dat_1=vartab_dat_i0.loc[vartab_dat_i0['udip1'].isin(list_i0_dat_1)]
UDI_list_i0_dat_1=UDI_i0_dat_1['UDI'].tolist()
#for r in list_i0_dat_1:
 #   vartab_dat_i0
#UDI_list_i0_dat_1
UDI_i0_dat_1['UDI'].to_csv('aUDI_i0_dat_1.csv',header='UDI')

UDI_i0_dat_1=pd.read_csv('aUDI_i0_dat_1.csv')
UDI_i0_dat_1
UDI_list_i0_dat_1=UDI_i0_dat_1['UDI'].tolist()
#UDI_list_i0_dat_1



In [ ]:
### All time variables: 


os.chdir("/temp_project/ukbb/data/i0")

vartab_tim_i0=vartab_time[vartab_time['udip2m']==0]
to_tim_list=freq_chk(vartab_tim_i0,'udip1',output=False)

### At i=0, lsit of time var only 1 measurement
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_tim_1=to_tim_list[to_tim_list['count']==1].index.tolist()

### At i=0, lsit of time var >= 2 measurements
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_tim_2=to_tim_list[to_tim_list['count']>1].index.tolist()
vartab_tim_i0.head()



list_i0_tim_1
#list_i0_tim_2   ###### Need to discuss about how to re-structure

UDI_i0_tim_1=vartab_tim_i0.loc[vartab_tim_i0['udip1'].isin(list_i0_tim_1)]
UDI_list_i0_tim_1=UDI_i0_tim_1['UDI'].tolist()
#for r in list_i0_tim_1:
 #   vartab_tim_i0
#UDI_list_i0_tim_1
UDI_i0_tim_1['UDI'].to_csv('aUDI_i0_tim_1.csv',header='UDI')

UDI_i0_tim_1=pd.read_csv('aUDI_i0_tim_1.csv')
UDI_i0_tim_1
UDI_list_i0_tim_1=UDI_i0_tim_1['UDI'].tolist()
#UDI_list_i0_tim_1



In [ ]:
### All curv variables: 


os.chdir("/temp_project/ukbb/data/i0")

vartab_cur_i0=vartab_curv[vartab_curv['udip2m']==0]
to_cur_list=freq_chk(vartab_cur_i0,'udip1',output=False)

### At i=0, lsit of curv var only 1 measurement
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_cur_1=to_cur_list[to_cur_list['count']==1].index.tolist()

### At i=0, lsit of curv var >= 2 measurements
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_cur_2=to_cur_list[to_cur_list['count']>1].index.tolist()
vartab_cur_i0.head()



list_i0_cur_1
#list_i0_cur_2   ###### Need to discuss about how to re-structure

UDI_i0_cur_1=vartab_cur_i0.loc[vartab_cur_i0['udip1'].isin(list_i0_cur_1)]
UDI_list_i0_cur_1=UDI_i0_cur_1['UDI'].tolist()
#for r in list_i0_cur_1:
 #   vartab_cur_i0
#UDI_list_i0_cur_1
UDI_i0_cur_1['UDI'].to_csv('aUDI_i0_cur_1.csv',header='UDI')

UDI_i0_cur_1=pd.read_csv('aUDI_i0_cur_1.csv')
UDI_i0_cur_1
UDI_list_i0_cur_1=UDI_i0_cur_1['UDI'].tolist()
#UDI_list_i0_cur_1



In [ ]:
### All text variables: 


os.chdir("/temp_project/ukbb/data/i0")

vartab_tex_i0=vartab_text[vartab_text['udip2m']==0]
to_tex_list=freq_chk(vartab_tex_i0,'udip1',output=False)

### At i=0, lsit of text var only 1 measurement
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_tex_1=to_tex_list[to_tex_list['count']==1].index.tolist()

### At i=0, lsit of text var >= 2 measurements
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_tex_2=to_tex_list[to_tex_list['count']>1].index.tolist()
vartab_tex_i0.head()



list_i0_tex_1
#list_i0_tex_2   ###### Need to discuss about how to re-structure

UDI_i0_tex_1=vartab_tex_i0.loc[vartab_tex_i0['udip1'].isin(list_i0_tex_1)]
UDI_list_i0_tex_1=UDI_i0_tex_1['UDI'].tolist()
#for r in list_i0_tex_1:
 #   vartab_tex_i0
#UDI_list_i0_tex_1
UDI_i0_tex_1['UDI'].to_csv('aUDI_i0_tex_1.csv',header='UDI')

UDI_i0_tex_1=pd.read_csv('aUDI_i0_tex_1.csv')
UDI_i0_tex_1
UDI_list_i0_tex_1=UDI_i0_tex_1['UDI'].tolist()
#UDI_list_i0_tex_1



In [ ]:
### All categorical (single) variables: 


os.chdir("/temp_project/ukbb/data/i0")

vartab_cas_i0=vartab_cat_si[vartab_cat_si['udip2m']==0]
vartab_cas_i0.CatLv.describe()
#freq_chk(vartab_cas_i0,'udip2n')

#freq_chk(vartab_cas_i0[vartab_cas_i0['udip2n']>0], )



os.chdir("/temp_project/ukbb/data/i0")

vartab_cas_i0=vartab_cat_si[vartab_cat_si['udip2m']==0]
to_cas_list=freq_chk(vartab_cas_i0,'udip1',output=False)
#to_cas_list

In [ ]:
os.chdir("/temp_project/ukbb/data/i0")

vartab_cas_i0=vartab_cat_si[vartab_cat_si['udip2m']==0]
vartab_cas_i0.CatLv.describe()
#freq_chk(vartab_cas_i0,'udip2n')

#freq_chk(vartab_cas_i0[vartab_cas_i0['udip2n']>0], )



os.chdir("/temp_project/ukbb/data/i0")

vartab_cas_i0=vartab_cat_si[vartab_cat_si['udip2m']==0]
to_cas_list=freq_chk(vartab_cas_i0,'udip1',output=False)

### At i=0, lsit of text var only 1 measurement
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_tex_1=to_tex_list[to_tex_list['count']==1].index.tolist()

### At i=0, lsit of text var >= 2 measurements
#tolist[tolist['count']==1].iloc[:,-1].tolist()
list_i0_tex_2=to_tex_list[to_tex_list['count']>1].index.tolist()
vartab_tex_i0.head()



list_i0_tex_1
#list_i0_tex_2   ###### Need to discuss about how to re-structure

UDI_i0_tex_1=vartab_tex_i0.loc[vartab_tex_i0['udip1'].isin(list_i0_tex_1)]
UDI_list_i0_tex_1=UDI_i0_tex_1['UDI'].tolist()
#for r in list_i0_tex_1:
 #   vartab_tex_i0
#UDI_list_i0_tex_1
UDI_i0_tex_1['UDI'].to_csv('aUDI_i0_tex_1.csv',header='UDI')

UDI_i0_tex_1=pd.read_csv('aUDI_i0_tex_1.csv')
UDI_i0_tex_1
UDI_list_i0_tex_1=UDI_i0_tex_1['UDI'].tolist()
#UDI_list_i0_tex_1


In [ ]:
vartab_cas_i0.head()